### PR TITLE
Category Section

### LINKED ISSUE #45: Category Section
Add Category Section below the Carousel based on the yelp app design.
WhatsApp Image 2023-04-27 at 10 37 57 AM

### LINKED ISSUE #50: Integrate Categories Api
Integrate Categories Yelp Fusion Api into yelp app.

[Api Link](https://api.yelp.com/v3/categories\)
[Documentation Link](https://docs.developer.yelp.com/reference/v3_all_categories)

### PR DESCRIPTION
**Pull request by Naveen-gali**
### What does this PR do?

- Adds Category Section

### Related Issues

- closes #45 
- closes #50 

### How to test

- Clone the Repository.
- Switch to feat/categorySection branch.
- Do yarn to install the dependencies.
- Do yarn run **build-stage-ios** for ios or **yarn run build-stage-android** for android.
- You Can see the Grid layout of the Categories on the screen

### Android Screen Recording.

https://github.com/Naveen-gali/Yelp-App/assets/122969303/efbe1710-521e-4c1f-8104-1715a126531b

### Ios Screen Recording.
https://github.com/Naveen-gali/Yelp-App/assets/122969303/7437a0be-da4a-45ae-9812-4a9807a87abb 

### Checklist
- [x] I have addressed all the requirements.
- [x] I have tested this code in both IOS and Android.
- [x] I have checked the code for formatting, syntax, and other best practices.
- [x] I have ensured that my changes do not break any existing functionality.
- [x] I have attached the necessary ScreenShots or Screen Recordings.
- [x] I have self reviewed the PR.

### Additional Notes 

- [FlatList Render Item Arrow Functions](https://stackoverflow.com/a/69732518)

### CODE REVIEW

File: src/components/CategoryItem/CategoryItem.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import {StyleSheet, Text, TouchableOpacity} from 'react-native';
+import {LocaleUtils, horizontalScale, verticalScale} from '../../utils';
+import {CategoryItemProps} from './CategoryItem.types';
+import {SvgWithCssUri} from 'react-native-svg';
```
Comment (prajnab-mm) [inline src/components/CategoryItem/CategoryItem.tsx]: format imports

File: src/components/CategoryItem/CategoryItem.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        {LocaleUtils.localizedText(title)}
+      </Text>
+    </TouchableOpacity>
+  );
+};
+
+const styles = StyleSheet.create({
+  iconContainer: {
+    flex: 1,
+    marginHorizontal: horizontalScale(14),
```
Comment (prajnab-mm) [inline src/components/CategoryItem/CategoryItem.tsx]: As mentioned earlier, reusable component should not have margins at the outer level

File: src/components/CategoryItem/CategoryItem.stories.ts
Comment (prajnab-mm) [inline src/components/CategoryItem/CategoryItem.stories.ts:1]: Add stories

File: src/screens/HomeScreen/components/CategorySection/CategorySection.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import {FlatList, ListRenderItemInfo, StyleSheet, View} from 'react-native';
+import {FeaturedCategoriesData} from '../../../../assets/data';
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/components/CategorySection/CategorySection.tsx]: Should import from assets and not assets/data

File: src/screens/HomeScreen/components/CategorySection/CategorySection.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
```
Comment (prajnab-mm) [inline src/screens/HomeScreen/components/CategorySection/CategorySection.tsx:2]: Add stories for every new component

File: src/assets/data/FeaturedCategoriesData.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
+[
+  {
```
Comment (prajnab-mm) [inline src/assets/data/FeaturedCategoriesData.json]: Do we not have API call for even this?

Comment (prajnab-mm) [reply src/assets/data/FeaturedCategoriesData.json]: Please list out the APIs we have that we can integrate in our app

File: src/screens/HomeScreen/components/CategorySection/CategorySection.tsx (hunk 4 above)
Comment (Naveen-gali) [reply src/screens/HomeScreen/components/CategorySection/CategorySection.tsx:2]: Added

File: src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      category.parent_aliases.includes(alias),
+    );
+  };
+
+  const renderMainContent = () => {
+    if (alias === 'more') {
+      return (
+        <FlatList
+          data={getData()}
+          renderItem={e => renderItem(e, true)}
```
Comment (prajnab-mm) [inline src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx:93]: Avoid adding arrow functions for renderItem

File: src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    );
+  };
+
+  const renderItem = (
+    renderItemProps: ListRenderItemInfo<CategoryInterface>,
+    showIcon: boolean,
+  ) => {
+    const {item, index} = renderItemProps;
+
+    const onPressHandler = () => {
```
Comment (prajnab-mm) [inline src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx]: Avoid functions within functions

File: src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+          {color: colors.text},
+        ]}
+        onPress={onPressHandler}
+        showIcon={showIcon}
+        showArrow={checkSubCategories(item.alias)}
+      />
+    );
+  };
+
+  const getData = () => {
```
Comment (prajnab-mm) [inline src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx]: Keep all render functions below and these above

File: src/utils/CategoryUtils/CategoryUtils.ts [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    'nightlife',
+    'active',
+    'beautysvc',
+    'homeservices',
+  ];
+
+  const moreItemArray = [
+    {
+      alias: 'more',
+      title: 'More',
```
Comment (prajnab-mm) [inline src/utils/CategoryUtils/CategoryUtils.ts]: Localisation

File: src/utils/CategoryUtils/CategoryUtils.ts [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import {CategoryInterface} from '../../models';
+import {FeaturedCategories, MoreCategories} from './CategoryUtils.types';
+
+const featuredCategories = (categories: CategoryInterface[]) => {
+  const topFeaturedItems = [
+    'restaurants',
+    'shopping',
+    'nightlife',
+    'active',
+    'beautysvc',
```
Comment (prajnab-mm) [inline src/utils/CategoryUtils/CategoryUtils.ts:14]: Add missing item

File: src/utils/CategoryUtils/CategoryUtils.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  ];
+
+  const featuredCategoryItems = categories.filter(category =>
+    topFeaturedItems.includes(category.alias),
+  );
+
+  const extraCategoryItems = categories
+    .filter(category => category.parent_aliases.length === 0)
+    .filter(category => !featuredCategoryItems.includes(category))
+    .slice(0, 7 - featuredCategoryItems.length);
```
Comment (prajnab-mm) [inline src/utils/CategoryUtils/CategoryUtils.ts]: Create const for the magic number 7 explaining what it is

File: src/utils/CategoryUtils/CategoryUtils.ts (hunk 10 above)
Comment (Naveen-gali) [reply src/utils/CategoryUtils/CategoryUtils.ts:14]: Added

File: src/utils/CategoryUtils/CategoryUtils.ts (hunk 11 above)
Comment (Naveen-gali) [reply src/utils/CategoryUtils/CategoryUtils.ts]: Added Constants

File: src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx (hunk 8 above)
Comment (Naveen-gali) [reply src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx]: Fixed.

File: src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx (hunk 7 above)
Comment (Naveen-gali) [reply src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx]: Replaced with useCallbacks by seeing this answer https://stackoverflow.com/a/69732518

File: src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx (hunk 6 above)
Comment (Naveen-gali) [reply src/screens/CategoryDetailScreen/CategoryDetailScreen.tsx:93]: Replaced with useCallbacks.

File: src/utils/CategoryUtils/CategoryUtils.ts (hunk 9 above)
Comment (Naveen-gali) [reply src/utils/CategoryUtils/CategoryUtils.ts]: All the other featured categories are coming only in english, keeping this one item in arabic i think doesnot look good so i havent changed it.

**Review by prajnab-mm — APPROVED** [review]

**Pull request merged**